rebuild/src/public_key.cc

```diff
@@ -0,0 +1,375 @@
+}
+
+const blst::P1 *PublicKey::AsJacobian()
+{
+    if (!_has_jacobian && !_has_affine)
+    {
+        throw Napi::Error::New(BlstBase::_env, "PublicKey not initialized");
+    }
+    if (!_has_jacobian)
+    {
+        _jacobian.reset(new blst::P1{_affine->to_jacobian()});
+        _has_jacobian = true;
+    }
+    return _jacobian.get();
+}
+
+const blst::P1_Affine *PublicKey::AsAffine()
+{
+    if (!_has_jacobian && !_has_affine)
+    {
+        throw Napi::Error::New(BlstBase::_env, "PublicKey not initialized");
+    }
+    if (!_has_affine)
+    {
+        _affine.reset(new blst::P1_Affine{_jacobian->to_affine()});
+        _has_affine = true;
+    }
+    return _affine.get();
+}
+
+bool PublicKey::NativeValidate()
+{
+    if (_has_jacobian && !_jacobian->is_inf() && _jacobian->in_group())
+    {
+        return true;
+    }
+    else if (_has_affine && !_affine->is_inf() && _affine->in_group())
+    {
+        return true;
+    }
+    return false;
+}
+
+/**
+ *
+ *
+ * PublicKeyArg
+ *
+ *
+ */
+PublicKeyArg::PublicKeyArg(Napi::Env env)
+    : BlstBase{env},
+      _public_key{nullptr},
+      _bytes{_env} {};
+
+PublicKeyArg::PublicKeyArg(Napi::Env env, Napi::Value raw_arg)
+    : PublicKeyArg{env}
+{
+    Napi::HandleScope scope(_env);
+    if (raw_arg.IsTypedArray())
+    {
+        _bytes = Uint8ArrayArg{_env, raw_arg, "PublicKeyArg"};
+        _bytes.ValidateLength(
+            _module->_public_key_compressed_length,
+            _module->_public_key_uncompressed_length);
+        if (_bytes.HasError())
+        {
+            // goto set_error to get more specific error message
+            // instead of Uint8Array error message
+            goto set_error;
+        }
+        Napi::Object wrapped = _module->_public_key_ctr.New({Napi::External<void *>::New(_env, nullptr)});
+        wrapped.TypeTag(&_module->_public_key_tag);
+        _ref = Napi::Persistent(wrapped);
+        _public_key = PublicKey::Unwrap(wrapped);
+        if (_bytes.IsZeroBytes(_bytes.Data(), 0, _bytes.ByteLength()))
+        {
+            _public_key->_jacobian.reset(new blst::P1{});
+            _public_key->_is_zero_key = true;
+        }
+        else
+        {
+            try
+            {
+                _public_key->_jacobian.reset(new blst::P1{_bytes.Data(), _bytes.ByteLength()});
+            }
+            catch (blst::BLST_ERROR &err)
+            {
+                std::ostringstream msg;
+                msg << _module->GetBlstErrorString(err) << ": Invalid PublicKey";
+                SetError(msg.str());
+                return;
+            }
+        }
+        _public_key->_has_jacobian = true;
+        return;
+    }
+    if (raw_arg.IsObject())
+    {
+        Napi::Object raw_obj = raw_arg.As<Napi::Object>();
+        if (raw_obj.CheckTypeTag(&_module->_public_key_tag))
+        {
+            _ref = Napi::Persistent(raw_obj);
+            _public_key = PublicKey::Unwrap(raw_obj);
+            return;
+        }
+    }
+
+set_error:
```

Comment:
not in favor of goto at all unless you are writing a stack implementation

goto is hard to prove correctness